server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java

```diff
@@ -104,17 +106,28 @@ public TransportAction(
             this.metadataCreateDataStreamService = metadataCreateDataStreamService;
             this.autoCreateIndex = autoCreateIndex;
             executor = (currentState, tasks) -> {
-                ClusterTasksResult.Builder<AckedClusterStateUpdateTask> builder = ClusterTasksResult.builder();
+                ClusterTasksResult.Builder<CreateIndexTask> builder = ClusterTasksResult.builder();
                 ClusterState state = currentState;
-                for (AckedClusterStateUpdateTask task : tasks) {
+                final Map<CreateIndexRequest, CreateIndexTask> successfulRequests = new HashMap<>(tasks.size());
+                for (CreateIndexTask task : tasks) {
                     try {
-                        state = task.execute(state);
+                        final CreateIndexTask successfulBefore = successfulRequests.putIfAbsent(task.request, task);
+                        if (successfulBefore == null) {
+                            state = task.execute(state);
+                        } else {
+                            // TODO: clean this up to just deduplicate the task listener instead of setting the generated name from
+                            // duplicate tasks here and then waiting for shards to become available multiple times in parallel for
+                            // each duplicate task
+                            task.indexNameRef.set(successfulBefore.indexNameRef.get());
+                        }
                         builder.success(task);
                     } catch (Exception e) {
                         builder.failure(task, e);
                     }
                 }
-                state = allocationService.reroute(state, "auto-create");
+                if (state != currentState) {
```

Comment:
small optimization in case the full batch doesn't do anything no need to reroute.

Reply:
I'm not sure about that. This adds one more CS update if no reroutes are pending and saves a reroute every now and then. A reroute is pretty fast these days though. With all the fixes we added recently I'm not sure it's worth it in 8.1.

Reply:
Hmm it's not _trivially_ fast in a large cluster still so I think we still prefer batching. Yes it's one more publication in an otherwise-quiet cluster but that's not really where the problems arise.